src/prog_models/prognostics_model.py

```diff
@@ -1201,58 +1204,96 @@ def estimate_params(self, runs : List[tuple] = None, keys : List[str] = None, ti
 
         Keyword Args:
-            keys (list[str]): 
+            keys (list[str]):
                 Parameter keys to optimize
             times (list[float]):
                 Array of times for each sample
             inputs (list[InputContainer]):
                 Array of input containers where input[x] corresponds to time[x]
             outputs (list[OutputContainer]):
                 Array of output containers where output[x] corresponds to time[x]
-            method (str, optional): 
+            method (str, optional):
                 Optimization method- see scipy.optimize.minimize for options
-            bounds (tuple or dict): 
+            bounds (tuple or dict, optional):
                 Bounds for optimization in format ((lower1, upper1), (lower2, upper2), ...) or {key1: (lower1, upper1), key2: (lower2, upper2), ...}
-            options (dict): 
+            options (dict, optional):
                 Options passed to optimizer. see scipy.optimize.minimize for options
-            runs (list[tuple], depreciated): 
+            runs (list[tuple], depreciated):
                 data from all runs, where runs[0] is the data from run 0. Each run consists of a tuple of arrays of times, input dicts, and output dicts. Use inputs, outputs, states, times, etc. instead
 
         See: examples.param_est
         """
-        from scipy.optimize import minimize
+        from scipy.optimize import minimize, show_options
 
         if keys is None:
             # if no keys provided, use all
             keys = [key for key in self.parameters.keys() if isinstance(self.parameters[key], Number)]
+        
+        for key in keys:
+            if key not in self.parameters:
+                raise ValueError("Passing in key that does not exist.")
 
         config = {
             'method': 'nelder-mead',
             'bounds': tuple((-np.inf, np.inf) for _ in keys),
-            'options': {'xatol': 1e-8},
+            'options': {'xatol': 1e-8}
         }
         config.update(kwargs)
 
+        if isinstance(times, set) or isinstance(inputs, set) or isinstance(outputs, set):
+            raise TypeError(f"Times, Inputs, and Outputs cannot be a Set. Sets are unordered by definition, so passing in arguments as Sets may have undefined behavior.")
+
+        # if parameters not in parent wrapper sequence, then place them into one.
+        if not runs and times and inputs and outputs:
+            if not isinstance(times[0], (Sequence, np.ndarray)):
+                times = [times]
+            if not isinstance(inputs[0], (Sequence, np.ndarray)):
+                inputs = [inputs]
+            if not isinstance(outputs[0], (Sequence, np.ndarray)):
+                outputs = [outputs]
+
         if runs is None and (times is None or inputs is None or outputs is None):
-            raise ValueError("Must provide either runs or times, inputs, and outputs")
+            if times is None and inputs is None and outputs is None:
```

Comment:
Question for @teubert - this is a very comprehensive check of all possible mistakes in passing times, inputs, and outputs (good job @aqitya). And while I think it's great, we did go from 1 if statement to 8. Computationally, these are probably decently fast, but is it worth it to include all so that we have very specific error message or should we save the computation time and print a less-specific error?

Reply:
I'm not too worried about the computation time, since it's in the outer loop. But it is a bit verbose. What if we do something like this:

```python
if runs is None:
	# If depreciated feature runs is not provided (will be removed in future version)

	# Check if required times, inputs, and outputs are present
	missing_args = [arg for arg in ('times', 'inputs', 'outputs) if globals()[arg] is None]
	if len(missing_inputs) > 0:
		# Concat into string
		missing_args_str = ', '.join(missing_args)
		missing_args_str = missing_args_str[:-2] # Remove last comma and space
		raise ValueError(f"Missing keyword arguments {missing_args_str}"
	
	# Check lengths of args
	if len(times) != len(inputs) ...
```

What do you think?